### PR TITLE
feat: new features for docking widget and a new color pallet

### PR DESCRIPTION
**Pull request by rohit2p**
**Purpose of PR?**:

This PR contain new features for docking widget.
such as: line thickness, line style, line transparency, new colour pallet for other widgets too, select/deselect buttons, restoration of previous states for the new ui, instead of hide/show groupbox disable the button which make more scene for the UI, automatic different better colour selection instead of only blue for the flightrack

### CODE REVIEW

File: mslib/msui/qt5/ui_multiple_flightpath_dockwidget.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 
-# Form implementation generated from reading ui file 'mslib/msui/ui/ui_multiple_flightpath_dockwidget.ui'
+# Form implementation generated from reading ui file '/home/zandar/Desktop/MSS/mslib/msui/ui/ui_multiple_flightpath_dockwidget.ui'
```
Comment (ReimarBauer) [inline mslib/msui/qt5/ui_multiple_flightpath_dockwidget.py]: normally there is a relative path only, have you used some options?

File: mslib/msui/qt5/ui_multiple_flightpath_dockwidget.py [hunk 2]
```diff
@@ -114,9 +148,25 @@ def retranslateUi(self, MultipleViewWidget):
...
-        self.pushButton_color.setText(_translate("MultipleViewWidget", "Change Color"))
+        self.groupBox.setTitle(_translate("MultipleViewWidget", "Flight Track style options"))
+        self.pushButton_color.setText(_translate("MultipleViewWidget", "Change Color of flight track"))
+        self.label.setText(_translate("MultipleViewWidget", "Line thnickness"))
+        self.label_3.setText(_translate("MultipleViewWidget", "Line style"))
+        self.label_2.setText(_translate("MultipleViewWidget", "Transparency"))
         self.labelStatus.setText(_translate("MultipleViewWidget", "Status: "))
+
+
+if __name__ == "__main__":
```
Comment (ReimarBauer) [inline mslib/msui/qt5/ui_multiple_flightpath_dockwidget.py]: usually there is no main section, was this an option used?

Comment (rohit2p) [reply mslib/msui/qt5/ui_multiple_flightpath_dockwidget.py]: Ik I'll clean up

**Comment by ReimarBauer** [comment]
There could be a spacer adden below the Line thickness, in the options menu, this looks to condensed with the frame surrounding  the secotion. 
image

**Comment by ReimarBauer** [comment]
I changed first with options the line thickness, then opened the docking widget. In the docking widget it remains its old value. 
image

ok, my fault, I selected a different operation.

good improvement, that is only changeable over the options.

image

**Comment by ReimarBauer** [comment]
But when I then change the color to e.g. red in the options

```
Fatal error in MSS 9.0.0 on Linux-6.5.0-10043-tuxedo-x86_64-with-glibc2.35
Python 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:40:35) [GCC 12.3.0]

Please report bugs in MSS to https://github.com/Open-MSS/MSS

Information about the fatal error:

Traceback (most recent call last):
  File "/home/reimar/PycharmProjects/2024/rohit2p/MSS/mslib/msui/multiple_flightpath_dockwidget.py", line 268, in ft_vertices_color
    self.dict_flighttrack[self.active_flight_track]["color"] = color
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: None

```


https://github.com/Open-MSS/MSS/assets/278728/ab826257-0b8e-4555-9908-586c51ca9ddb


changing thickness gives no failure

**Comment by crolf** [comment]
Thanks for your nice progress. 
I found something confusing, when switching the active flight path. The line style is changing. Is it by purpose ? I guess that there need to be an additional sync for the non active flight path:

[Bildschirmaufzeichnung vom 11.07.2024, 09:08:07.webm](https://github.com/Open-MSS/MSS/assets/75527994/917803c6-92b8-4271-b958-9c8d3bc4e892)

**Comment by gisi90** [comment]
Works nicely for the non-active flight track and transparency option is great. 
Because the active flight track cannot be edited, the line properties options are hidden. However, when the non-active flight track is changed to active, the line properties options remain visible but non of the settings take effect. Only for changing color I get the Infomessage that the color of the active flight track can only be changed in under "options". Maybe, there is a way to hide the line properties options again when a flight track is made active.
https://github.com/Open-MSS/MSS/assets/54586339/7d5e7f87-8010-4d33-bb5c-264a38a28f05

File: mslib/msui/qt5/ui_topview_mapappearance.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 
-# Form implementation generated from reading ui file '/home/zandar/Desktop/MSS/mslib/msui/ui/ui_topview_mapappearance.ui'
+# Form implementation generated from reading ui file 'ui_topview_mapappearance.ui'
```
Comment (ReimarBauer) [inline mslib/msui/qt5/ui_topview_mapappearance.py]: where is the ui file? please add it to.

Comment (rohit2p) [reply mslib/msui/qt5/ui_topview_mapappearance.py]: i think i have already pushed that in this commit "[select feature and a ui](https://github.com/Open-MSS/MSS/pull/2428/commits/56a31d6d2252ef83e0084546d71cc788d8200bdb)"

File: mslib/msui/multiple_flightpath_dockwidget.py [hunk 4]
```diff
@@ -103,6 +114,58 @@ def remove(self):
         self.map.ax.figure.canvas.draw()
 
 
+class CustomColorDialog(QtWidgets.QDialog):
```
Comment (ReimarBauer) [inline mslib/msui/multiple_flightpath_dockwidget.py]: this should be moved to mslib.utils.colordialog and then imported.

When you create that new module use our template header.

File: mslib/msui/sideview.py [hunk 5]
```diff
@@ -44,6 +44,58 @@
 WMS = 0
 
 
+class CustomColorDialog(QtWidgets.QDialog):
+    """
```
Comment (ReimarBauer) [inline mslib/msui/sideview.py]: import from mslib.utils.colordialog

File: mslib/msui/topview.py [hunk 6]
```diff
@@ -56,6 +57,58 @@
 MULTIPLEFLIGHTPATH = 5
 
 
+class CustomColorDialog(QtWidgets.QDialog):
+    """
```
Comment (ReimarBauer) [inline mslib/msui/topview.py]: import from mslib.utils.colordialog

File: mslib/msui/multiple_flightpath_dockwidget.py [hunk 7]
```diff
@@ -398,6 +517,75 @@ def set_linewidth(self):
         else:
             self.labelStatus.setText("Status: No flight track selected")
 
+    def set_transparency(self):
+        """
+        Change the line transparency of the selected flight track.
```
Comment (ReimarBauer) [inline mslib/msui/multiple_flightpath_dockwidget.py:484]: is this doubled with line 1039?

**Review by ReimarBauer — CHANGES_REQUESTED** [review]
see comments.

File: mslib/msui/multiple_flightpath_dockwidget.py (hunk 7 above)
Comment (rohit2p) [reply mslib/msui/multiple_flightpath_dockwidget.py:484]: no the other method is for mscolab

**Comment by ReimarBauer** [comment]
the test_setColour needs to be changed to mock the new Color widget

**Comment by matrss** [comment]
The CI test timeouts are what I already suspected and have to do with you replacing the color dialog. After a test run there is a lingering color dialog open that just isn't closed properly, so the test run hangs:

Screenshot from 2024-08-08 15-34-43

The test that fails is this one:
```
FAILED tests/_test_msui/test_sideview.py::Test_MSS_SV_OptionsDialog::test_setColour - AssertionError: assert 0 == 1
```

**Comment by rohit2p** [comment]
> The CI test timeouts are what I already suspected and have to do with you replacing the color dialog. After a test run there is a lingering color dialog open that just isn't closed properly, so the test run hangs:
> 

> The test that fails is this one:
> 
> ```
> FAILED tests/_test_msui/test_sideview.py::Test_MSS_SV_OptionsDialog::test_setColour - AssertionError: assert 0 == 1
> ```

I was thinking of fixing it when i am writing test code. I need a better understanding of testing i guess

**Comment by ReimarBauer** [comment]
The legend in the picture is not yet part of the feature, I added it to illustrate what I want to have.  It is overlayed by pqiv  for this example.

When we look at 


image1

v1 is activated and the user has decided by the options that it has the color blue and is a solid line

The example operation is a dashed line.

Now the user wants to keep all colors and styles because they are refered in communications and activates the example operation

image2

by the activation this "example" line now gets blue and solid. And the previous v1 is now green. It should stay blue.


What I want to have, is that the colors not change without user interaction. 
At the moment a different flightrack is activated the color used before have to be kept and changed in the "..Vertices Line Color and the option to that color. Currently they get overwritten by what was defined in options.

In the example this would mean if I now again would click v1, the color of the operation flight track which I can change starts with green,

**Comment by ReimarBauer** [comment]
When all operations are selected to be shown and you toggle activations of operations one by one get removed from the view. At the end only one is shown. 

selection_drawing

**Comment by ReimarBauer** [comment]
The wording maybe changed:

Left: Select all Flighttracks
Right: Select all Operations

You maybe want also to use same layer as for the list widgets to align them by the size.

tabs

**Comment by ReimarBauer** [comment]
> > The CI test timeouts are what I already suspected and have to do with you replacing the color dialog. After a test run there is a lingering color dialog open that just isn't closed properly, so the test run hangs:
> 
> > The test that fails is this one:
> > ```
> > FAILED tests/_test_msui/test_sideview.py::Test_MSS_SV_OptionsDialog::test_setColour - AssertionError: assert 0 == 1
> > ```
> 
> I was thinking of fixing it when i am writing test code. I need a better understanding of testing i guess

You can mock the CustomColorDialog.exec_

We likly merge only without failing tests. 

once getColor added instead of many on_color_selected maybe also an option.

**Comment by ReimarBauer** [comment]
> The wording maybe changed:
> 
> Left: Select all Flighttracks Right: Select all Operations
> 
> You maybe want also to use same layer as for the list widgets to align them by the size.
> 

This should maybe arranged similiar to:

image

File: mslib/msui/multiple_flightpath_dockwidget.py [hunk 8]
```diff
@@ -298,15 +349,34 @@ def save_waypoint_model_data(self, wp_model, listWidget):
...
+        Get a random color from custom colors ensuring no repeats.
+        """
+        available_colors = [color for color in self.custom_colors if color not in self.used_colors]
+        if not available_colors:
+            # Reset the used colors if all colors have been used
+            self.used_colors = []
+            available_colors = self.custom_colors.copy()
+
+        selected_color = random.choice(available_colors)
+        self.used_colors.append(selected_color)
```
Comment (matrss) [inline mslib/msui/multiple_flightpath_dockwidget.py:366]: This (and the reset further above) modify a class variable. This means that appending something here will affect all other MultipleFlightPathControlWidget objects. This is what I mean, for reference:
```
>>> class A:
...     x = []
... 
>>> a1 = A()
>>> a2 = A()
>>> a1.x
[]
>>> a2.x
[]
>>> a1.x.append("test123")
>>> a1.x
['test123']
>>> a2.x
['test123']
```

Is this intentional or a bug?

File: mslib/msui/multiple_flightpath_dockwidget.py [hunk 9]
```diff
@@ -342,19 +423,25 @@ def select_color(self):
...
-                        self.dict_flighttrack[wp_model]["color"] = color.getRgbF()
-                        self.color_change = True
-                        self.list_flighttrack.currentItem().setIcon(self.show_color_icon(self.get_color(wp_model)))
-                        self.dict_flighttrack[wp_model]["patch"].update(color=self.dict_flighttrack[wp_model]["color"])
+                    color_dialog = CustomColorDialog(self)
+                    color_dialog.color_selected.connect(lambda color: self.apply_color(wp_model, color))
+                    color_dialog.exec_()
             else:
-                self.labelStatus.setText("Check Mark the flighttrack to change its color.")
+                self.labelStatus.setText("Status: Check Mark the flight track to change it's Colour.")
```
Comment (matrss) [inline mslib/msui/multiple_flightpath_dockwidget.py]: Not sure about "flighttrack" vs "flight track", but "its" without an apostrophe is correct. "color" vs "colour" is just american-influenced vs british-influenced english, and in any case should be lowercase.

File: mslib/msui/multiple_flightpath_dockwidget.py [hunk 10]
```diff
@@ -342,19 +423,25 @@ def select_color(self):
...
             self.labelStatus.setText("Status: No flight track selected")
 
+    def apply_color(self, wp_model, color):
+        if color.isValid():
+            self.dict_flighttrack[wp_model]["color"] = color.getRgbF()
+            self.color_change = True
+            self.list_flighttrack.currentItem().setIcon(self.show_color_icon(self.get_color(wp_model)))
+            self.dict_flighttrack[wp_model]["patch"].update(
+                color=self.dict_flighttrack[wp_model]["color"],
+                linewidth=self.dict_flighttrack[wp_model]["linewidth"])
```
Comment (matrss) [inline mslib/msui/multiple_flightpath_dockwidget.py]: Why does this require passing the linewidth? AFAICT it isn't changed.

File: mslib/msui/multiple_flightpath_dockwidget.py [hunk 11]
```diff
@@ -325,6 +395,17 @@ def create_list_item(self, wp_model):
...
+    def selectAll(self, state):
+        """
+        select/deselect local operations
+        """
+        for i in range(self.list_flighttrack.count()):
+            item = self.list_flighttrack.item(i)
+            if state == Qt.Checked:
+                item.setCheckState(Qt.Checked)
+            else:
+                item.setCheckState(Qt.Unchecked)
```
Comment (matrss) [inline mslib/msui/multiple_flightpath_dockwidget.py]:
```suggestion
            item.setCheckState(state)
```
?

File: mslib/msui/qt5/ui_multiple_flightpath_dockwidget.py (hunk 1 above)
Comment (matrss) [reply mslib/msui/qt5/ui_multiple_flightpath_dockwidget.py]: Same question

File: mslib/msui/multiple_flightpath_dockwidget.py (hunk 8 above)
Comment (rohit2p) [reply mslib/msui/multiple_flightpath_dockwidget.py:366]: yes you are right it should be an instance variable

**Comment by ReimarBauer** [comment]
Thx for the UI improvements, 

Because this box does the same for the left and the right list, I would use more "open" wording

image

e.g.: 
- Style:
- Change Color

**Comment by ReimarBauer** [comment]
speed improvement: 44 operations of different size in 21 seconds for me :)

tried also on an hardware server, there it needs only 3.57 seconds.

If someone is interested how I checked
[measure.patch](https://github.com/user-attachments/files/16612331/measure.txt)

**Comment by rohit2p** [comment]
> speed improvement: 44 operations of different size in 21 seconds for me :)
> 
> tried also on an hardware server, there it needs only 3.57 seconds.
> 
> If someone is interested how I checked [measure.patch](https://github.com/user-attachments/files/16612331/measure.txt)

Can you explain more i dont get it what u trying to say?

**Comment by ReimarBauer** [comment]
> > speed improvement: 44 operations of different size in 21 seconds for me :)
> > tried also on an hardware server, there it needs only 3.57 seconds.
> > If someone is interested how I checked [measure.patch](https://github.com/user-attachments/files/16612331/measure.txt)
> 
> Can you explain more i dont get it what u trying to say?

I examined the different servers I have with your code. Once your branch is merged we have on the UI a speed improvement. The next improvement must be done on server hardware or code. But it looks like our File I/O is a bit slow on the server.

**Comment by ReimarBauer** [comment]
When we have a flight track together with the operations and switch from flight track to an operation the marked flight track is not anymore on screen.

enabled_flight_track_gone

**Comment by rohit2p** [comment]
> When we have a flight track together with the operations and switch from flight track to an operation the marked flight track is not anymore on screen.
> 


That problem is also in develop branch. tho I don't have a idea rn, but if you want i can work on fixing it after i am done with regex and legend integration functionality

**Comment by ReimarBauer** [comment]
> The legend in the picture is not yet part of the feature, I added it to illustrate what I want to have. It is overlayed by pqiv for this example.
> 
> When we look at
> 
> image1
> 
> v1 is activated and the user has decided by the options that it has the color blue and is a solid line
> 
> The example operation is a dashed line.
> 
> Now the user wants to keep all colors and styles because they are refered in communications and activates the example operation
> 
> image2
> 
> by the activation this "example" line now gets blue and solid. And the previous v1 is now green. It should stay blue.
> 
> What I want to have, is that the colors not change without user interaction. At the moment a different flightrack is activated the color used before have to be kept and changed in the "..Vertices Line Color and the option to that color. Currently they get overwritten by what was defined in options.
> 
> In the example this would mean if I now again would click v1, the color of the operation flight track which I can change starts with green,

Talked with @joernu76 short about this. This needs a further discussion because someone might want a defined color always for the active one. We maybe need a different issue for that after implementation.

**Review by ReimarBauer — APPROVED** [review]
Talked with @joernu76 too. The PR brings some nice improvements with it. Further work will be carried out in a follow-up.

**Pull request merged**